src/gadgets/sumcheck.rs

```diff
@@ -0,0 +1,336 @@
+      AllocatedNum::<Scalar>::alloc(&mut cs, || Ok(eval.get_value().unwrap()))?;
+    let allocated_eval_aux: AllocatedNum<Scalar> =
+      AllocatedNum::<Scalar>::alloc(&mut cs, || Ok(eval_aux.get_value().unwrap()))?;
+    cs.enforce(
+      || format!("eval + coeff = eval_aux{}", i),
+      |lc| lc + allocated_eval.get_variable() + coeff.get_variable(),
+      |lc| lc + CS::one(),
+      |lc| lc + allocated_eval_aux.get_variable(),
+    );
+    eval = eval_aux;
+  }
+  let allocated_eval: AllocatedNum<Scalar> =
+    AllocatedNum::<Scalar>::alloc(&mut cs, || Ok(eval.get_value().unwrap()))?;
+
+  Ok(allocated_eval)
+}
+
+// method compatible with src/spartan/sumcheck.rs > UniPoly::evaluate()
+pub fn uni_evaluate<Scalar, CS>(
+  mut cs: CS,
+  coeffs: Vec<AllocatedNum<Scalar>>,
+  r: AllocatedNum<Scalar>,
+) -> Result<AllocatedNum<Scalar>, SynthesisError>
+where
+  Scalar: PrimeField + PrimeFieldBits,
+  CS: ConstraintSystem<Scalar>,
+{
+  let mut eval: Num<Scalar> = Num::<Scalar>::from(coeffs[0].clone());
+  let mut power = r.clone();
+  for (i, coeff) in coeffs.iter().enumerate().skip(1) {
+    // logic: eval_aux = eval + power * coeff
+    let eval_aux = eval.clone().add(&Num::<Scalar>::from(
+      power.mul(cs.namespace(|| format!("mul {}", i)), &coeff)?,
+    ));
+    // r1cs constr: power * coeff = eval - eval_aux
+    let allocated_eval: AllocatedNum<Scalar> =
+      AllocatedNum::<Scalar>::alloc(&mut cs, || Ok(eval.get_value().unwrap()))?;
+    let allocated_eval_aux: AllocatedNum<Scalar> =
+      AllocatedNum::<Scalar>::alloc(&mut cs, || Ok(eval_aux.get_value().unwrap()))?;
+    cs.enforce(
+      || format!("power * coeff = eval_aux - eval{}", i),
+      |lc| lc + power.get_variable(),
+      |lc| lc + coeff.get_variable(),
+      |lc| lc + allocated_eval_aux.get_variable() - allocated_eval.get_variable(),
+    );
+    eval = eval_aux;
+
+    // logic: power_aux = power * r
+    let power_aux = power.mul(cs.namespace(|| format!("mul2 {}", i)), &r)?;
+    // r1cs constr: power_aux = power * r
+    cs.enforce(
+      || format!("power_aux = power * r {}", i),
+      |lc| lc + power.get_variable(),
+      |lc| lc + r.get_variable(),
+      |lc| lc + power_aux.get_variable(),
+    );
+    power = power_aux;
+  }
+  let allocated_eval: AllocatedNum<Scalar> =
+    AllocatedNum::<Scalar>::alloc(&mut cs, || Ok(eval.get_value().unwrap()))?;
+
+  Ok(allocated_eval)
+}
+
+#[cfg(test)]
+mod tests {
+  use super::*;
+  use crate::bellperson::{
+    r1cs::{NovaShape, NovaWitness},
+    {shape_cs::ShapeCS, solver::SatisfyingAssignment},
+  };
+  use crate::spartan::sumcheck::UniPoly;
+  use pasta_curves::{arithmetic::CurveAffine, pallas, pallas::Scalar, vesta};
+
+  use crate::traits::{Group, TranscriptEngineTrait, TranscriptReprTrait};
+  type PastaG1 = pasta_curves::pallas::Point;
+  use crate::spartan::polynomial::MultilinearPolynomial;
+  use crate::spartan::sumcheck::SumcheckProof;
+
+  fn synthetize_uni_evaluate<Scalar, CS>(
```

Comment:
What is the role of this function? Are the commented stuff needed?

Reply:
The role is the same as in other `synthesize_` methods in the tests of other gadgets such as 
https://github.com/privacy-scaling-explorations/Nova/tree/b53229d07c50cd135006b319995bd29799cfc2e7/src/gadgets/ecc.rs#L1020

The commented stuff will not be needed, leaving it for the moment meanwhile the PR is in draft mode, in order to check stuff while the PR is in progress.

Note about typo: in the sumcheck.rs code the word `synthetize` should be `synthesize` as in the rest of methods with that word in the other Nova files, I'll update this in the next commit.